File: Btree/Btree/btree.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <climits>

struct node {
  int data;
  node *left;
  node *right;
  node(int data) {
    this->data = data;
    left = nullptr;
    right = nullptr;
  }
};

class btree {
private:
  node *root;
public:

  void add_node(int data, node *&root) {
    if (root == nullptr) {
      root = new node(data);
    } else {
      node* curr = root;
      if (data < curr->data) {
        if (curr->left == nullptr) {
          curr->left = new node(data);
        } else {
          add_node(data, curr->left);
        }
      } else {
        if (curr->right == nullptr) {
          curr->right = new node(data);
        } else {
          add_node(data, curr->right);
        }
      }
    }
  }

  int get_data(int data, node *&root) {
    if (root == nullptr){
      return INT_MAX; 
    }     
    if (root->data == data){
      return root->data; 
    }
    node *curr = root;
      if (data < curr->data) {
      return get_data(data, curr->left);
      } else {
      return get_data(data, curr->right);
      }
    }
  

  void pre_order(node *root) {
    if (root == nullptr) {
      return;
    }
    std::cout << root->data;
    pre_order(root->left);
    pre_order(root->right);
  }

  void in_order(node *root) {
    if (root == nullptr) {
      return;
    } else {
      in_order(root->left);
      std::cout << root->data;
      in_order(root->right);
    }
  }

  void post_order(node *root) {
    if (root == nullptr) {
      return;
    } else {
      post_order(root->left);
      post_order(root->right);
      std::cout << root->data;
    }
  }

  void iterative_pre_order(node *root) {
    std::stack<node *> s;
    node *curr = root;
    while (curr != nullptr || !s.empty()) {
      if (curr != nullptr) {
        std::cout << curr->data;
        s.push(curr);
        curr = curr->left;
      } else {
        node *prev = s.top();
        s.pop();
        curr = prev->right;
      }
    }
  }

  void iterative_in_order(node *root) {
    std::stack<node *> s;
    node *curr = root;
    while (curr != nullptr || !s.empty()) {
      if (curr != nullptr) {
        s.push(curr);
        curr = curr->left;
      } else {
        curr = s.top();
        s.pop();
        std::cout << curr->data;
        curr = curr->right;
      }
    }
  }

  void bfs(node *root) {
    std::queue<node *> q;
    node *curr = root;
    q.push(curr);
    while (!q.empty()) {
      node *curr = q.front();
      if (curr->left != nullptr) {
        q.push(curr->left);
      }
      if (curr->right != nullptr) {
        q.push(curr->right);
      }
      q.pop();
    }
  }
};

File: Btree/Btree/Btree.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

/*
 * subtree(x) = (x) and its descendants 
 * depth(x) = # ancesotrs = # edged in path from (x) to root
 * height(x) = # edges in longest downward path from (x) = max depth in subtree(x) 
 * height of tree = max depth 
 */
public class Btree {
    public class Node {
        public int data;
        public Node left;
        public Node right;
        public Node parent;

        public Node(int data) {
            this.data = data;
        }
    }

    public Node root;

    public void addNode(int data, Node current) {
        if (root == null) {
            root = new Node(data);
        } else if (data < current.data) {
            if (current.left == null) {
                current.left = new Node(data);
                current.left.parent = current; 
            } else {
                addNode(data, current.left);
            }
        } else {
            if (current.right == null) {
                current.right = new Node(data);
                current.right.parent = current; 
            } else {
                addNode(data, current.right);
            }
        }
    }

    public int getData(int data, Node current) {
        if (current == null || current.data == data) {
            return current.data;
        } else if (data < current.data) {
            return getData(data, current.left);
        } else {
            return getData(data, current.right);
        }
    }

    public void preOrder(Node current) {
        if (current == null) {
            return; 
        } else {
            System.out.print(current.data + " ");
            preOrder(current.left);
            preOrder(current.right);
        }
    }

    public void inOrder(Node current){
        if (current == null){
            return; 
        }
        else {
            inOrder(current.left); 
            System.out.print(current.data + " ");
            inOrder(current.right); 
        }
    }

    public void postOrder(Node current){
        if (current == null){
            return;  
        }
        postOrder(current.left); 
        postOrder(current.right); 
        System.out.print(current.data + " "); 

    }

    public void nonRecInOrder(Node current){
        Stack <Node> stack = new Stack<>(); 
        while(current != null || !stack.isEmpty()){
            if (current != null){
                stack.push(current); 
                current = current.left; 
            } else {
                current = stack.pop(); 
                System.out.print(current.data + " ");
                current = current.right; 

            }
        }
    }

    public int minTree(Node current){
        while(current.left != null){
            current = current.left; 
        }
        return current.data; 
    }

    public int maxTree(Node current){
        while(current.right != null){
            current = current.right; 
        }
        return current.data; 
    }

    public int recMaxTree(Node current){
        if (current.right == null){
            return current.data; 
        }
        else return recMaxTree(current.right); 
    }

    public int recMinTree(Node current){
        if (current.left == null){
            return current.data; 
        }
        else return recMinTree(current.left); 
    }

    public int getSuccessor(Node current){
        if (current.right != null){
            return minTree(current.right);  
        }   else {
                Node temp = current.parent; 
                while(temp != null && temp.data < current.data){
                temp = temp.parent; 
            }
            return temp.data; 
        }
    }

    public void levelOrder(Node root) {
        if (root == null) {
            return;
        }
        Queue<Node> Q = new LinkedList<>();
        Q.add(root);
        while (!Q.isEmpty()) {
            Node current = Q.peek();
            System.out.print(current.data + " ");
            if (current.left != null) {
                Q.add(current.left);
            }
            if (current.right != null) {
                Q.add(current.right);
            }
            Q.poll();
        }
    }

    public int countLevels(Node root){
        if (root == null){
            return 0; 
        }
        int count = 0; 
        Queue<Node> Q = new LinkedList<>();
        Q.add(root); 
        while(!Q.isEmpty()){
            int nodeCount = Q.size(); 
            while (nodeCount > 0){
                Node current = Q.poll(); 
                if (current.left != null){
                    Q.add(current.left); 
                }
                if (current.right != null){
                    Q.add(current.right); 
                }
                --nodeCount; 
             }
        ++count;
        }
        return count; 
    }



 

    /*public int getPredecessor(Node current){
        if (current == root){ 
            return current.data; 
        }    
        if (current.left != null){
            return recMaxTree(current.left); 
         }  else if (return recMa
    }*/

    public static void main(String[] args) {
        Btree tree = new Btree();
        tree.addNode(5, tree.root);
        tree.addNode(3, tree.root);
        tree.addNode(7, tree.root);
        tree.addNode(2, tree.root);
        tree.addNode(4, tree.root);
        tree.addNode(6, tree.root);
        tree.addNode(8, tree.root);

        System.out.println("Preorder Traversal: ");
        tree.preOrder(tree.root);
        System.out.println("\nInorder Traversal: ");
        tree.inOrder(tree.root);
        System.out.println("\nPostorder Traversal: ");
        tree.postOrder(tree.root);

        System.out.println("\nNon-recursive Inorder Traversal: ");
        tree.nonRecInOrder(tree.root);

        System.out.println("\nLevel Order Traversal: ");
        tree.levelOrder(tree.root);

        System.out.println("\nMinimum Value: " + tree.recMinTree(tree.root));
        System.out.println("Maximum Value: " + tree.recMaxTree(tree.root));

        System.out.println("Successor of 3: " + tree.getSuccessor(tree.root.left));

        System.out.println("Number of levels in the tree: " + tree.countLevels(tree.root));
    }
}
